_source/_posts/2023-09-15-java-spring-boot-microservices.md

```diff
@@ -0,0 +1,722 @@
+
+The good news is I figured out a workaround! If you pass a valid `audience` parameter to Auth0, you'll get a JWT for the access token. I logged an [issue to improve the Okta Spring Boot starter](https://github.com/okta/okta-spring-boot/issues/596) and added a `SecurityConfiguration` class to solve the problem in the meantime.
+
+```java
+package com.example.apigateway.config;
+
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.security.config.web.server.ServerHttpSecurity;
+import org.springframework.security.oauth2.client.registration.ReactiveClientRegistrationRepository;
+import org.springframework.security.oauth2.client.web.server.DefaultServerOAuth2AuthorizationRequestResolver;
+import org.springframework.security.oauth2.client.web.server.ServerOAuth2AuthorizationRequestResolver;
+import org.springframework.security.oauth2.core.endpoint.OAuth2AuthorizationRequest;
+import org.springframework.security.web.server.SecurityWebFilterChain;
+
+import java.util.function.Consumer;
+
+@Configuration
+public class SecurityConfiguration {
+
+    @Value("${okta.oauth2.audience:}")
+    private String audience;
+
+    private final ReactiveClientRegistrationRepository clientRegistrationRepository;
+
+    public SecurityConfiguration(ReactiveClientRegistrationRepository clientRegistrationRepository) {
+        this.clientRegistrationRepository = clientRegistrationRepository;
+    }
+
+    @Bean
+    public SecurityWebFilterChain filterChain(ServerHttpSecurity http) throws Exception {
+        http
+            .authorizeExchange(authz -> authz
+                .anyExchange().authenticated()
+            )
+            .oauth2Login(oauth2 -> oauth2
+                .authorizationRequestResolver(
+                    authorizationRequestResolver(this.clientRegistrationRepository)
+                )
+            );
+        return http.build();
+    }
+
+    private ServerOAuth2AuthorizationRequestResolver authorizationRequestResolver(
+        ReactiveClientRegistrationRepository clientRegistrationRepository) {
+
+        var authorizationRequestResolver =
+            new DefaultServerOAuth2AuthorizationRequestResolver(clientRegistrationRepository);
+        authorizationRequestResolver.setAuthorizationRequestCustomizer(
+            authorizationRequestCustomizer());
+
+        return authorizationRequestResolver;
+    }
+
+    private Consumer<OAuth2AuthorizationRequest.Builder> authorizationRequestCustomizer() {
+        return customizer -> customizer
+            .additionalParameters(params -> params.put("audience", audience));
+    }
+}
+```
+
+I was able to get the OpenFeign client to work by adding a couple of properties to enable OAuth:
```

Comment:
I understood you removed feign?

Reply:
Yes. This is a weird entry, I'll admit. I figured out how to get OpenFeign working with OAuth before I got Spring Cloud Gateway working. Should I move it up or eliminate it altogether? I can also just mention it and link to the branch. https://github.com/oktadev/auth0-java-microservices-examples/pull/2

Once Spring Cloud Gateway has MVC support, I plan to create a new example in the repo and update this post.

Reply:
Not sure if I understand the phrase, is it with or without `spring-cloud-starter-gateway`? Maybe mention it at the end of the section, before the refresh tokens subtitle, and link to the branch as you said. I would say (if correct) "I was able to get the initial `api-gateway` version (with Spring Cloud OpenFeign) working with OAuth ....

Reply:
I added your suggestion. However, I might've misunderstood. Are you saying I should move to to its own section. If so, that makes sense. Once the MVC version of Spring Cloud Gateway comes out (in Spring Boot 3.2), I plan to update this post and have two sections for MVC and for WebFlux.

Reply:
imo this might just confuse people so probably better to just remove now and add later when using boot 3.2

Reply:
I'll comment it out so I remember how to do it.

Reply:
Commented out in https://github.com/oktadev/okta-blog/pull/1449/commits/808e453bd7a6f49d63c8254c89040651fa781fd6.